Mathlib/Data/Vector.lean

```diff
@@ -0,0 +1,257 @@
+@[match_pattern]
+def cons : α → Vector α n → Vector α (Nat.succ n)
+  | a, ⟨v, h⟩ => ⟨a :: v, congrArg Nat.succ h⟩
+#align vector.cons Vector.cons
+
+
+/-- The length of a vector -/
+@[reducible, nolint unusedArguments]
+def length (_ : Vector α n) : ℕ :=
+  n
+#align vector.length Vector.length
+
+open Nat
+
+/-- the first element of a vector with length at least `1` -/
+def head : Vector α (Nat.succ n) → α
+  | ⟨[], h⟩ => by contradiction
+  | ⟨a :: _, _⟩ => a
+#align vector.head Vector.head
+
+/-- the head of a vector obtained by prepending is the element prepended -/
+theorem head_cons (a : α) : ∀ v : Vector α n, head (cons a v) = a
+  | ⟨_, _⟩ => rfl
+#align vector.head_cons Vector.head_cons
+
+/-- the tail of a vector, with an empty vector having empty tail  -/
+def tail : Vector α n → Vector α (n - 1)
+  | ⟨[], h⟩ => ⟨[], congrArg pred h⟩
+  | ⟨_ :: v, h⟩ => ⟨v, congrArg pred h⟩
+#align vector.tail Vector.tail
+
+/-- the tail of a vector obtained by prepending is the vector prepended to -/
+theorem tail_cons (a : α) : ∀ v : Vector α n, tail (cons a v) = v
+  | ⟨_, _⟩ => rfl
+#align vector.tail_cons Vector.tail_cons
+
+/-- prepending the head of a vector to its tail gives the vector -/
+@[simp]
+theorem cons_head_tail : ∀ v : Vector α (succ n), cons (head v) (tail v) = v
+  | ⟨[], h⟩ => by contradiction
+  | ⟨a :: v, h⟩ => rfl
+#align vector.cons_head_tail Vector.cons_head_tail
+
+/-- list obtained from a vector -/
+def toList (v : Vector α n) : List α :=
+  v.1
+#align vector.to_list Vector.toList
+
+/-- nth element of a vector, indexed by a `Fin` type -/
+def nth : ∀ _ : Vector α n, Fin n → α
+  | ⟨l, h⟩, i => l.nthLe i.1 (by rw [h] ; exact i.2)
+#align vector.nth Vector.nth
+
+/-- appending a vector to another -/
+def append {n m : Nat} : Vector α n → Vector α m → Vector α (n + m)
+  | ⟨l₁, h₁⟩, ⟨l₂, h₂⟩ => ⟨l₁ ++ l₂, by simp [*]⟩
+#align vector.append Vector.append
+
+/- warning: vector.elim -> Vector.elim is a dubious translation:
+lean 3 declaration is
+  forall {α : Type.{u_1}} {C : forall {n : ℕ},
+  (Vector.{u_1} α n) -> Sort.{u}}, (forall (l : List.{u_1} α),
+  C (List.length.{u_1} α l) (Subtype.mk.{succ u_1} (List.{u_1} α)
+  (fun (l_1 : List.{u_1} α) => Eq.{1} ℕ (List.length.{u_1} α l_1)
+  (List.length.{u_1} α l)) l (Vector.Elim._proof_1.{u_1} α l))) ->
+  (forall {n : ℕ} (v : Vector.{u_1} α n), C n v)
+but is expected to have type
+  forall {α : Type.{_aux_param_0}} {C : forall {n : ℕ}, (Vector.{_aux_param_0} α n) -> Sort.{u}},
+  (forall (l : List.{_aux_param_0} α),
+  C (List.length.{_aux_param_0} α l) (Subtype.mk.{succ _aux_param_0} (List.{_aux_param_0} α)
+  (fun (l_1 : List.{_aux_param_0} α) => Eq.{1} ℕ (List.length.{_aux_param_0} α l_1)
+  (List.length.{_aux_param_0} α l)) l (rfl.{1} ℕ (List.length.{_aux_param_0} α l)))) ->
+  (forall {n : ℕ} (v : Vector.{_aux_param_0} α n), C n v)
+Case conversion may be inaccurate. Consider using '#align vector.elim Vector.elimₓ'. -/
+/-- elimination rule for `Vector` -/
```

Comment:
Hmm. At the very least a porting note should be made here; it seems that in mathlib3 there is some kind of `vector.elim._proof_1` mentioned in the definition, and in Lean 4 it was replaced by a `rfl`. Hopefully this won't cause any problems though.

Reply:
Where should the porting note go? I kept the comment which I believe was made by @gebner in mathlib3port.